File: data_structures/doubly_linked_list.go
package data_structures

import "fmt"

type nodeDoublyList struct {
	Value    int
	Next     *nodeDoublyList
	Previous *nodeDoublyList
}

type DoublyLinkedList struct {
	Head *nodeDoublyList
	Tail *nodeDoublyList
}

func (d *DoublyLinkedList) Prepend(value int) {
	newNode := &nodeDoublyList{Value: value}

	if d.Head == nil {
		d.Head = newNode
		d.Tail = newNode
	} else {
		newNode.Next = d.Head
		d.Head.Previous = newNode
		d.Head = newNode
	}
}

func (d *DoublyLinkedList) Append(value int) {
	newNode := &nodeDoublyList{Value: value}

	if d.Head == nil {
		d.Head = newNode
		d.Tail = newNode
	} else {
		newNode.Previous = d.Tail
		d.Tail.Next = newNode
		d.Tail = newNode
	}
}

func (d *DoublyLinkedList) InsertAt(value, index int) {
	newNode := &nodeDoublyList{Value: value}

	if index <= 0 || d.Head == nil {
		d.Prepend(value)
		return
	}

	current := d.Head
	for i := 0; i < index-1 && current.Next != nil; i++ {
		current = current.Next
	}

	if current.Next == nil {
		d.Append(value)
		return
	}

	newNode.Next = current.Next
	newNode.Previous = current
	current.Next.Previous = newNode
	current.Next = newNode
}

func (d *DoublyLinkedList) RemoveFirst() *nodeDoublyList {
	if d.Head == nil {
		return nil
	}

	temp := d.Head
	d.Head = d.Head.Next

	if d.Head == nil {
		d.Tail = nil
	} else {
		d.Head.Previous = nil
	}

	temp.Next = nil // clean removed node
	return temp
}

func (d *DoublyLinkedList) RemoveLast() *nodeDoublyList {
	if d.Head == nil {
		return nil
	}

	temp := d.Tail
	d.Tail = d.Tail.Previous

	if d.Tail == nil {
		d.Head = nil
	} else {
		d.Tail.Next = nil
	}

	temp.Previous = nil // clean removed node
	return temp
}

func (d *DoublyLinkedList) RemoveAt(index int) *nodeDoublyList {
	if d.Head == nil || index < 0 {
		return nil
	}

	if index == 0 {
		return d.RemoveFirst()
	}

	current := d.Head
	for i := 0; i < index && current != nil; i++ {
		current = current.Next
	}

	if current == nil {
		return nil
	}

	if current.Next == nil {
		return d.RemoveLast()
	}

	current.Previous.Next = current.Previous
	current.Next.Previous = current.Previous

	current.Previous = nil
	current.Next = nil
	return current
}

func (d *DoublyLinkedList) Get(index int) *nodeDoublyList {
	if index < 0 {
		return nil
	}

	current := d.Head
	for i := 0; i < index && current != nil; i++ {
		current = current.Next
	}

	return current
}

func (d *DoublyLinkedList) Print() {
	current := d.Head
	for current != nil {
		fmt.Print(current.Value)
		if current.Next != nil {
			fmt.Print(" <-> ")
		}
		current = current.Next
	}

	fmt.Println()
}

File: c/data-structures/doubly-linked-list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *next;
    struct Node *previous;
} Node;

typedef struct doubly_linked_list {
    struct Node *head;
    struct Node *tail;
} doubly_linked_list;

Node *create_node(int data) {
    Node *new_node = (Node *)malloc(sizeof(Node));

    if (new_node == NULL) {
        fprintf(stderr, "Error in allocating memory\n");
        exit(EXIT_FAILURE);
    }

    new_node->data = data;
    new_node->next = NULL;
    new_node->previous = NULL;

    return new_node;
}

void prepend(doubly_linked_list *list, int data) {
    Node *new_node = create_node(data);
    new_node->next = list->head;
    list->head = new_node;
}

void print_list(doubly_linked_list *list) {
    Node *current = list->head;

    while (current != NULL) {
        printf("%d -> ", current->data);

        current = current->next;
    }

    printf("NULL\n");
}

int main() {
    doubly_linked_list my_list = {NULL};

    prepend(&my_list, 10);
    prepend(&my_list, 20);
    prepend(&my_list, 30);
    prepend(&my_list, 40);
    prepend(&my_list, 50);

    print_list(&my_list);

    return 0;
}
